### PR TITLE
WSL Stuck on Import/Refresh

### LINKED ISSUE #1285: Project import or "Refresh SAP Commerce Project" stuck on "Scanning SAP Commerce Projects"
### Environment Details
- **IntelliJ IDEA Version**: IntelliJ IDEA 2024.3.1
- **SAP Commerce IntelliJ Plugin Version**: 2024.3.2
- **Operating System**: Ubuntu 24.04 (via WSL2 on Windows 11)
- **Java Version**: Oracle JDK 11
- **SAP Commerce Version**: 2105
- Log from importing: [idea.log](https://github.com/user-attachments/files/18084822/idea.log)

The import process gets stuck indefinitely on "Scanning SAP Commerce Projects."
- Tried clearing caches and restarting IntelliJ with disabling all other plugins but faced the same issue.
image

### PR DESCRIPTION
**Pull request by mburakgurler**
WSL Stuck on Import/Refresh
* added isPathInWSLDistribution function for checking if importing project directory is in WSL
* with isPathInWSLDistribution if function return true scanSubdirectories will use wsl function for it will use Files.list instead of Files.newDirectoryStream

Signed-off-by: Burak Gürler gurlerburak9@gmail.com

### CODE REVIEW

File: src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java [hunk 1]
```diff
@@ -723,14 +727,23 @@ private void findModuleRoots(
...
-            moduleRootMap,
-            excludedFromScanning,
-            acceptOnlyHybrisModules,
-            rootProjectDirectory.toPath(),
-            progressListenerProcessor
-        );
+        scanForSubdirectories(moduleRootMap, excludedFromScanning, acceptOnlyHybrisModules, rootProjectDirectory.toPath(), progressListenerProcessor);
+    }
 
+    private void scanForSubdirectories(final Map<DIRECTORY_TYPE, Set<File>> moduleRootMap, final Set<File> excludedFromScanning, final boolean acceptOnlyHybrisModules, final Path rootProjectDirectory, final TaskProgressProcessor<File> progressListenerProcessor) {
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java]: both exceptions should be caught by `DefaultHybrisProjectDescriptor.setRootDirectoryAndScanForModules`, can you please comment the reason to catch them here?

File: src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java [hunk 2]
```diff
@@ -764,6 +777,45 @@ private void scanSubdirectories(
...
+    private void scanSubdirectoriesWSL(@NotNull final Map<DIRECTORY_TYPE, Set<File>> moduleRootMap, final Set<File> excludedFromScanning, final boolean acceptOnlyHybrisModules, @NotNull final Path rootProjectDirectory, @Nullable final TaskProgressProcessor<File> progressListenerProcessor) throws InterruptedException, IOException {
+        if (!Files.isDirectory(rootProjectDirectory)) {
+            return;
+        }
+
+        try (final Stream<Path> stream = Files.list(rootProjectDirectory)) {
+            stream.filter(file -> {
+                try {
+                    return Objects.nonNull(file) && Files.isDirectory(file) && !isDirectoryExcluded(file) && (!Files.isSymbolicLink(file) || followSymlink);
+                } catch (Exception e) {
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java]: what can throw an `Exception` here? looks like nothing.

File: src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java [hunk 3]
```diff
@@ -764,6 +777,45 @@ private void scanSubdirectories(
         }
     }
 
+    private void scanSubdirectoriesWSL(@NotNull final Map<DIRECTORY_TYPE, Set<File>> moduleRootMap, final Set<File> excludedFromScanning, final boolean acceptOnlyHybrisModules, @NotNull final Path rootProjectDirectory, @Nullable final TaskProgressProcessor<File> progressListenerProcessor) throws InterruptedException, IOException {
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java:772]: method style can be slightly simplified:
```
private void scanSubdirectoriesWSL(@NotNull final Map<DIRECTORY_TYPE, Set<File>> moduleRootMap, final Set<File> excludedFromScanning, final boolean acceptOnlyHybrisModules, @NotNull final Path rootProjectDirectory, @Nullable final TaskProgressProcessor<File> progressListenerProcessor) throws InterruptedException, IOException {
        if (!Files.isDirectory(rootProjectDirectory)) return;

        try (final Stream<Path> stream = Files.list(rootProjectDirectory)) {
            final var moduleRoots = stream
                .filter(Objects::nonNull)
                .filter(Files::isDirectory)
                .filter(Predicate.not(this::isDirectoryExcluded))
                .filter(file -> !Files.isSymbolicLink(file) || followSymlink)
                .map(Path::toFile)
                .toList();
            for (final var moduleRoot : moduleRoots) {
                findModuleRoots(moduleRootMap, excludedFromScanning, acceptOnlyHybrisModules, moduleRoot, progressListenerProcessor);
            }
        }
    }
```

File: src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java [hunk 4]
```diff
@@ -764,6 +777,45 @@ private void scanSubdirectories(
...
+                    findModuleRoots(moduleRootMap, excludedFromScanning, acceptOnlyHybrisModules, file.toFile(), progressListenerProcessor);
+                } catch (IOException | InterruptedException e) {
+                    LOG.error("Error while processing directories", e);
+                }
+            });
+        }
+    }
+
+
+    public static boolean isPathInWSLDistribution(@NotNull final Path rootProjectDirectory) {
```
Comment (mlytvyn) [inline src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java:789]: what may throw an exception?

alternative style:
```
public static boolean isPathInWSLDistribution(@NotNull final Path rootProjectDirectory) {
        return WslDistributionManager.getInstance().getInstalledDistributions().stream()
            .map(WSLDistribution::getUNCRootPath)
            .map(String::valueOf)
            .filter(StringUtils::isNoneBlank)
            .anyMatch(wslRoot -> rootProjectDirectory.toString().startsWith(wslRoot));
    }
```

**Review by mlytvyn — COMMENTED** [review]
@mburakgurler , thank you for your contribution!

File: src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java (hunk 1 above)
Comment (mburakgurler) [reply src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java]: Yeah you're right I should not handle exceptions here, muscle memory I think,
Try Catch blocks cleaned and signatures added to scanForSubdirectories

File: src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java (hunk 3 above)
Comment (mburakgurler) [reply src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java:772]: Done :)

File: src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java (hunk 4 above)
Comment (mburakgurler) [reply src/com/intellij/idea/plugin/hybris/project/descriptors/DefaultHybrisProjectDescriptor.java:789]: Alternative style applied.

**Comment by mburakgurler** [comment]
@mlytvyn Code is refactored as you wanted, also tested in debug and work Idea.

**Review by mlytvyn — APPROVED** [review]

**Comment by mlytvyn** [comment]
@mburakgurler , thank you for your contribution and so long waiting feature by many!

**Comment by darkweizer** [comment]
Thank you, thank you, thank you !! My project works again. I searched for hours where the problem could come from

**Pull request merged**